[cmake/win32] fixes for building with abnormal configuration

Kwiboo opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
<!--- Describe your change in detail -->
Building fails (on windows) when version tag is empty, app name is changed or when optical/upnp/airtunes is disabled

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
Fixes building of abnormal configuration on windows

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
Compile tested with cmake 3.6.3 and visual studio 2015 on windows 10

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Kwiboo committed:
[win32] fix build when optical is disabled

Kwiboo committed:
[win32] fix linking when airtunes or upnp is disabled

Kwiboo committed:
[cmake] fix build error when version tag is empty

Kwiboo committed:
[cmake] use app name target for application manifest

Kwiboo committed:
[win32] remove unused has audio define

hudokkow commented:
Thanks for looking into this. CMake changes look sane. As for the rest, we need an win dev.

ping @Paxxi 
jenkins build this please

Razzeee commented:
jenkins build failed last time with this PR.
Let's do it again and see why.
jenkins build this please

Kwiboo commented:
If it helps any review of this PR:
`HAS_AIRTUNES` and `HAS_UPNP` removed from `xbmc/system.h` is defined earlier in the same file depending on cmake config and should not cause any changes in defines with normal cmake config.
`HAS_AUDIO` was removed because I could not find it used anywhere inside Kodi: https://github.com/xbmc/xbmc/search?q=HAS_AUDIO

hudokkow commented:
Anyone has a reason not to merge this? If so, speak in the next 24h or I'll push it in.
@afedchin, do you have time for a quick look?

afedchin commented:
what should I take a look on? I know nothing about cmake.

hudokkow commented:
The ifdeferry? It's not a CMake invention... ;)

afedchin commented on xbmc/storage/windows/Win32StorageProvider.cpp:
```diff
@@ -96,6 +98,7 @@ CDetectDisc::CDetectDisc(const std::string &strPath, const bool bautorun)
 bool CDetectDisc::DoWork()
 {
   CLog::Log(LOGDEBUG, "%s: Optical media found in drive %s", __FUNCTION__, m_strPath.c_str());
+#ifdef HAS_DVD_DRIVE
```
this ifdef isn't needed because the ifdef above avoid running this code. if you want to keep this for safety, I think the log record above shouldn't be occurs also.

Kwiboo replied:
I'll update to ifdef away the complete `CDetectDisc` class.

`g_mediaManager.GetDiskLabel` and `g_mediaManager.GetDiskUniqueId` are the methods missing when `HAS_DVD_DRIVE` is undefined.

afedchin replied:
got it, but please move this up to avoid logging also.

Kwiboo commented:
Rebased and a moved `ifdef` in `CDetectDisc::DoWork()`

Completely removing `CDetectDisc` was a rather obtrusive change as it also required ifdef in https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/windows/WinEventsWin32.cpp#L790 so I just moved the ifdef in `CDetectDisc::DoWork()`

MartijnKaijser commented:
jenkins build this please
any further comments/objections?

Paxxi commented:
Looks good to me

hudokkow commented:
jenkins build and merge
